### PR TITLE
Hashtable update + dynamic growth

### PR DESCRIPTION
**Pull request by cgzones**

### CODE REVIEW

File: Hashtable.c [hunk 1]
```diff
@@ -79,10 +135,42 @@ void Hashtable_delete(Hashtable* this) {
...
+            walk->next = NULL;
+
+            unsigned int newIndex = walk->key % newSize;
+            HashtableItem** newBucketPtr = &newBuckets[newIndex];
+            while (*newBucketPtr)
+               newBucketPtr = &(*newBucketPtr)->next;
+            *newBucketPtr = walk;
+
+            walk = next;
+         }
```
Comment (BenBE) [inline Hashtable.c:161]: Split resizing into its own routine? Useful if you know you are adding quite a few items soon.

File: Hashtable.c [hunk 2]
```diff
@@ -94,10 +182,12 @@ void Hashtable_put(Hashtable* this, unsigned int key, void* value) {
          break;
       } else
          bucketPtr = &((*bucketPtr)->next);
```
Comment (BenBE) [inline Hashtable.c:184]: Care to embrace this branch?

**Review by BenBE — CHANGES_REQUESTED** [review]

**Review by BenBE — APPROVED** [review]
LGTM.

Let's stick with Eratosthenes for now; Miller-Rabin outperforms that only for n>500k, which we'll rarely hit.

File: Hashtable.c [hunk 3]
```diff
@@ -79,10 +136,49 @@ void Hashtable_delete(Hashtable* this) {
...
+
+   assert(Hashtable_isConsistent(this));
+}
+
+void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
+
+   assert(Hashtable_isConsistent(this));
+   assert(this->size > 0);
+
+   // grow on load > 1.0
```
Comment (fasterit) [inline Hashtable.c:175]: `load` is a well defined term for a tool that is used in systems monitoring, better use `fill level` 1/2

File: Hashtable.c [hunk 4]
```diff
@@ -120,11 +219,17 @@ void* Hashtable_remove(Hashtable* this, unsigned int key) {
          }
       }
    }
+
    assert(Hashtable_isConsistent(this));
+
+   // shrink on load < 0.2
```
Comment (fasterit) [inline Hashtable.c:225]: `load` is a well defined term for a tool that is used in systems monitoring, better use `fill level` 2/2

**Comment by fasterit** [comment]
8fc3a91 also has `load factor` in the commit message which is a bit misleading

Do we have any metrics that this MR improves something (beyond the obvious cleanups in the first commit)?
The whole prime sieving thing looks a bit academic to me. Which may be fine but academics prove the superiority of their ideas :-)

File: Hashtable.c (hunk 3 above)
Comment (BenBE) [reply Hashtable.c:175]: Load factor is a common term for describing hash tables too:
https://en.wikipedia.org/wiki/Hash_table#Dynamic_resizing

**Comment by BenBE** [comment]
Given that [load factor](https://en.wikipedia.org/wiki/Hash_table#Key_statistics) is default terminology for hash tables, I don't think so. It somewhat collides with the system load if shortened, but apart from this I don't think there's much confusion - in particular as the hash table doesn't do any system load monitoring, but it's own number of items.

Regarding the performance impact using prime sizes: There is empirical evidence (across several real-world implementations) that most keys used in computer science and common data sets have reduced probability of key collisions when using a size of a hashtable that is prime, while for powers of two this probability is much higher, due to the nature of most data sets.

One minor optimizations we might take is avoiding the prime calculations altogether and use a fixed list of (prime) sizes. That way we skip the prime testing stuff completely and for hashtables above 2^30 entries there's little chance we'll encounter them in normal program flow (and could limit to 1GB for backing storage in such situations anyway).

**Comment by fasterit** [comment]
It sure is default terminology in CS but load in htop means load, not the load factor of some hash table. Comments (and commit messages) should help not confuse.
We have the pid which is a nice unique key (as we don't care for re-used pids, if ever), I don't get the need for primes in our context.

**Comment by BenBE** [comment]
> I don't get the need for primes in our context.

The primes are not used as a key for entries in the hashtable, but as the size of the hashtable. Doing so allows to drastically reduce the situation where different keys (from two different processes) map to the same entry in the hashtable, and thus improving the performance of accessing data inside the hashtable (as the entry does not have to be searched for through additional lookups).

**Comment by fasterit** [comment]
> > I don't get the need for primes in our context.
> 
> The primes are not used as a key for entries in the hashtable, but as the size of the hashtable. Doing so allows to drastically reduce the situation where different keys (from two different processes) map to the same entry in the hashtable, and thus improving the performance of accessing data inside the hashtable (as the entry does not have to be searched for through additional lookups).

Sure, but we have pids. So 300 on a common desktop, 1000 on a server, 10000 on some busy number crunching AI superbrick.
I'm not sure we are solving a real problem for htop here.

**Comment by BenBE** [comment]
Processes aren't the only information we are storing in there. There's also open files/inode information (cf. #238) among other information. So having the Hashtable be able to grow dynamically if the need arises isn't wrong; and doesn't make the code any slowe when not used (apart from one negligible check for when to grow it). The prime stuff could even be hardcoded for sizes up to 10^9 entries and we'd safe loads of cycles there, if we really wanted.

So overall: I think it's worth the effort.

**Comment by cgzones** [comment]
Some performance data on `Hashtable_get()`:

with standard separate chaining (https://github.com/htop-dev/htop/pull/277/commits/7085a49c3340f52b2fd5d8441afcd395465baf1e):
```
Hashtable_get count=1000 time=54861ns avg_time=54ns
Hashtable_get count=1000 time=113638ns avg_time=113ns
Hashtable_get count=1000 time=131835ns avg_time=131ns
Hashtable_get count=1000 time=144952ns avg_time=144ns
Hashtable_get count=1000 time=141755ns avg_time=141ns
Hashtable_get count=1000 time=138190ns avg_time=138ns
Hashtable_get count=1000 time=144508ns avg_time=144ns
Hashtable_get count=1000 time=145610ns avg_time=145ns
Hashtable_get count=1000 time=143832ns avg_time=143ns
Hashtable_get count=1000 time=135880ns avg_time=135ns
Hashtable_get count=1000 time=147611ns avg_time=147ns
Hashtable_get count=1000 time=139903ns avg_time=139ns
Hashtable_get count=1000 time=145853ns avg_time=145ns
Hashtable_get count=1000 time=138972ns avg_time=138ns
Hashtable_get count=1000 time=160244ns avg_time=160ns
Hashtable_get count=1000 time=148218ns avg_time=148ns
Hashtable_get count=1000 time=142785ns avg_time=142ns
Hashtable_get count=1000 time=142553ns avg_time=142ns
```

with standard Linear probing (https://github.com/htop-dev/htop/pull/277/commits/acbb93e6e8bbc588b7114ab6f3589e109121e74f):
```
Hashtable_get count=1000 time=56499ns avg_time=56ns
Hashtable_get count=1000 time=63503ns avg_time=63ns
Hashtable_get count=1000 time=66080ns avg_time=66ns
Hashtable_get count=1000 time=87105ns avg_time=87ns
Hashtable_get count=1000 time=84611ns avg_time=84ns
Hashtable_get count=1000 time=81494ns avg_time=81ns
Hashtable_get count=1000 time=77085ns avg_time=77ns
Hashtable_get count=1000 time=83541ns avg_time=83ns
Hashtable_get count=1000 time=86132ns avg_time=86ns
Hashtable_get count=1000 time=76165ns avg_time=76ns
Hashtable_get count=1000 time=82045ns avg_time=82ns
Hashtable_get count=1000 time=87207ns avg_time=87ns
Hashtable_get count=1000 time=80570ns avg_time=80ns
Hashtable_get count=1000 time=78787ns avg_time=78ns
Hashtable_get count=1000 time=78137ns avg_time=78ns
Hashtable_get count=1000 time=83362ns avg_time=83ns
Hashtable_get count=1000 time=82325ns avg_time=82ns
```

with Robin Hood Linear probing (https://github.com/htop-dev/htop/pull/277/commits/69f9d6585b3140ffda35ae035e28e60516d82e6a):
```
Hashtable_get count=1000 time=59177ns avg_time=59ns
Hashtable_get count=1000 time=63398ns avg_time=63ns
Hashtable_get count=1000 time=64487ns avg_time=64ns
Hashtable_get count=1000 time=81746ns avg_time=81ns
Hashtable_get count=1000 time=87998ns avg_time=87ns
Hashtable_get count=1000 time=74290ns avg_time=74ns
Hashtable_get count=1000 time=79050ns avg_time=79ns
Hashtable_get count=1000 time=86503ns avg_time=86ns
Hashtable_get count=1000 time=69286ns avg_time=69ns
Hashtable_get count=1000 time=73422ns avg_time=73ns
Hashtable_get count=1000 time=77854ns avg_time=77ns
Hashtable_get count=1000 time=85680ns avg_time=85ns
Hashtable_get count=1000 time=79989ns avg_time=79ns
Hashtable_get count=1000 time=78102ns avg_time=78ns
Hashtable_get count=1000 time=90531ns avg_time=90ns
Hashtable_get count=1000 time=78936ns avg_time=78ns
Hashtable_get count=1000 time=81299ns avg_time=81ns
```

File: Hashtable.c [hunk 5]
```diff
@@ -98,71 +92,91 @@ static unsigned int nextPrime(unsigned int n) {
...
 void Hashtable_setSize(Hashtable* this, unsigned int size) {
 
    assert(Hashtable_isConsistent(this));
 
-   unsigned int newSize = nextPrime(size);
-   HashtableItem** newBuckets = (HashtableItem**) xCalloc(newSize, sizeof(HashtableItem*));
+   HashtableItem* oldBuckets = this->buckets;
+   unsigned int oldSize = this->size;
+
+   this->size = nextPrime(size);
```
Comment (BenBE) [inline Hashtable.c:167]: Should validate that the new size fits all existing items.

As far as I understand the changed implementation we no longer have a chain of items attached per bucket, thus the number of items is strictly limited by the number of buckets.

File: Hashtable.c [hunk 6]
```diff
@@ -171,87 +185,120 @@ void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
...
-            assert(Hashtable_isConsistent(this));
-            return NULL;
-         } else {
-            assert(Hashtable_isConsistent(this));
-            return value;
+   void* res = NULL;
+
+   while (this->buckets[index].value) {
+      if (this->buckets[index].key == key) {
+         res = this->owner ? (free(this->buckets[index].value), NULL) : this->buckets[index].value;
```
Comment (BenBE) [inline Hashtable.c:212]: Please write a full if statement. This is hard to read (although I know what you mean).

File: Hashtable.c [hunk 7]
```diff
@@ -171,87 +185,120 @@ void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
...
+   return res;
 }
 
+#include <time.h>
+
 void* Hashtable_get(Hashtable* this, hkey_t key) {
+
+   struct timespec pre;
+   clock_gettime(CLOCK_MONOTONIC, &pre);
+
```
Comment (BenBE) [inline Hashtable.c:255]: Debug code, please remove.

File: Hashtable.c [hunk 8]
```diff
@@ -171,87 +185,120 @@ void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
...
+   time += post.tv_nsec - pre.tv_nsec;
+
+   if (counter >= 1000) {
+      fprintf(stderr, "Hashtable_get count=%u time=%ldns avg_time=%ldns\n",
+              counter,
+              time,
+              time / counter);
+      counter = 0;
+      time = 0;
+   }
```
Comment (BenBE) [inline Hashtable.c:289]: Debug code, please remove.

File: Hashtable.c [hunk 9]
```diff
@@ -171,87 +185,120 @@ void Hashtable_put(Hashtable* this, hkey_t key, void* value) {
...
+
+      assert(index != origIndex);
    }
 
    assert(Hashtable_isConsistent(this));
+   assert(Hashtable_get(this, key) == NULL);
 
-   // shrink on load < 0.2
-   if (5 * this->items < this->size)
+   // shrink on load < 0.125
```
Comment (BenBE) [inline Hashtable.c:242]: Prefer `load factor` in comment to avoid confusion with `(system) load` from `htop` context.

**Review by BenBE — CHANGES_REQUESTED** [review]
Object access LGTM, but not tried in ASAN ;-)

**Review by BenBE — APPROVED** [review]

**Pull request merged**